### PR TITLE
grpc-json: Handle streaming backend error cases correctly

### LINKED ISSUE #8315: grpc-json: Handle streaming backend error cases correctly
The grpc-json-transcoder filter does not handle errors from **streaming** backends correctly. I opened this issue to document the current and proposed downstream responses.

Note that #8009 introduced a new `convert_grpc_status` flag to deduce the HTTP response code and body from the gRPC backend error. This flag introduces another edge case (below).

I think the logic for unary responses is OK. I've only seen these issues with streaming backends.

cc @qiwzhang 
cc @ascheglov 

## Streaming Error with flag set

Configuration: The backend returns a gRPC error via a trailer-only response. The filter is configured with `convert_grpc_status=true`.

#### Current Behavior

The filter returns a HTTP 200 OK. The grpc status flag is set in one of the response headers. Assuming that #8312 will be merged in, the response will be `[]` with content-type `application/json`.

#### Proposed Behavior

The filter should return a response that matches the unary case with `convert_grpc_status` set:
- The HTTP code should match the gRPC status code.
- The response body should be a JSON with the error message.

## Streaming Error without flag

Configuration: The backend returns a gRPC error via a trailer-only response. The filter is configured without `convert_grpc_status` (default behavior).

#### Current Behavior

Same as above: The filter returns a HTTP 200 OK. The grpc status flag is set in one of the response headers. Assuming that #8312 will be merged in, the response will be `[]` with content-type `application/json`.

#### Proposed Behavior

The filter should return a response that exposes the error:
- The HTTP code should match the gRPC status code.
- The body should be completely empty (""), not an empty JSON array
- Content type should not be set, or it shouldn't matter what it is set to

## Other Notes

I would prefer to get #8312 merged in first to fix the content type issue. Please advise on how to handle this, as it may be a breaking change (?)

### PR DESCRIPTION
**Pull request by ascheglov**
When a streaming backend returns an error in a trailer-only response,
with the 'convert_grpc_status' option enabled,
instead of writing the [] (empty array) in reply-body
transcode the error in the headers and reply with {"code":...}.

Risk Level: Low
Testing: added an integraion test, tested manually.
Documentation: n/a
Release notes: n/a
Fixes #8315

### CODE REVIEW

**Comment by ascheglov** [comment]
On second thought, this could be improved.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 1]
```diff
@@ -451,46 +451,50 @@ Http::FilterTrailersStatus JsonTranscoderFilter::encodeTrailers(Http::HeaderMap&
...
-  }
-
   // If there was no previous headers frame, this |trailers| map is our |response_headers_|,
   // so there is no need to copy headers from one to the other.
   bool is_trailers_only_response = response_headers_ == &trailers;
 
   const absl::optional<Grpc::Status::GrpcStatus> grpc_status =
       Grpc::Common::getGrpcStatus(trailers);
   bool status_converted_to_json = grpc_status && maybeConvertGrpcStatus(*grpc_status, trailers);
+  if (status_converted_to_json) {
```
Comment (qiwzhang) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: The code is very hard to follow with so many if-else.  Can we split them into two functions, one handling all logic of convert_to_json=true case, and another handling false case.  Not need to worry about sharing code between them.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 2]
```diff
@@ -592,8 +590,26 @@ bool JsonTranscoderFilter::maybeConvertGrpcStatus(Grpc::Status::GrpcStatus grpc_
     return false;
   }
 
+  response_headers_->Status()->value(Grpc::Utility::grpcToHttpStatus(grpc_status));
+
+  if (is_trailers_only_response) {
```
Comment (qiwzhang) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: not need to have this if.  non_trailer_only_response has been bailed out in line 560

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 3]
```diff
@@ -592,8 +590,26 @@ bool JsonTranscoderFilter::maybeConvertGrpcStatus(Grpc::Status::GrpcStatus grpc_
...
+    // Drop the gRPC status headers, we already have them in the JSON body.
+    response_headers_->removeGrpcStatus();
+    response_headers_->removeGrpcMessage();
+    response_headers_->remove(Http::Headers::get().GrpcStatusDetailsBin);
+  }
+
+  // remove Trailer headers if the client connection was http/1
+  if (encoder_callbacks_->streamInfo().protocol() != Http::Protocol::Http2) {
+    static const Http::LowerCaseString trailer_key = Http::LowerCaseString("trailer");
+    response_headers_->remove(trailer_key);
```
Comment (qiwzhang) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: do we need to remove "trailer" header for non_json_converting case?

Comment (ascheglov) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: yes, and it's already there, in `encodeTrailers`

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc (hunk 2 above)
Comment (ascheglov) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: true, however I'm planning to remove that `if (!is_trailers_only_response) return` in next PR, which would fix the http connection manager so that it would allow to add data from the encodeTrailers handler.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 4]
```diff
@@ -592,8 +590,26 @@ bool JsonTranscoderFilter::maybeConvertGrpcStatus(Grpc::Status::GrpcStatus grpc_
...
+  if (is_trailers_only_response) {
+    // Drop the gRPC status headers, we already have them in the JSON body.
+    response_headers_->removeGrpcStatus();
+    response_headers_->removeGrpcMessage();
+    response_headers_->remove(Http::Headers::get().GrpcStatusDetailsBin);
+  }
+
+  // remove Trailer headers if the client connection was http/1
+  if (encoder_callbacks_->streamInfo().protocol() != Http::Protocol::Http2) {
+    static const Http::LowerCaseString trailer_key = Http::LowerCaseString("trailer");
```
Comment (lizan) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: (I know this were here before) since you're here, can you make this pointer or extract to a method with CONSTRUCT_ON_FIRST_USE, so doesn't cause static-initialization fiasco?

**Comment by nareddyt** [comment]
#8312 was merged, could you ensure your changes still work as expected with the new tests?

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc (hunk 4 above)
Comment (ascheglov) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: done

**Comment by ascheglov** [comment]
> #8312 was merged, could you ensure your changes still work as expected with the new tests?

I've checked it and there is no need to change anything.

For streaming responses, the "content-type" header will be set twice, but that's alright because `setReference` is cheap.

**Comment by ascheglov** [comment]
Apparently the asan build is stuck. How do I restart it?

**Comment by nareddyt** [comment]
> Apparently the asan build is stuck. How do I restart it?

Push an empty commit: https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#triggering-ci-re-run-without-making-changes

**Comment by lizan** [comment]
Can you merge master? You commit doesn't contain asan job config in azp.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc (hunk 2 above)
Comment (lizan) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: does that mean you're going to handle status to json conversion in non trailers_only_response case?

Comment (ascheglov) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: Yes, I'll remove this in #8404 .

We can transcode status into json as long as upstream doesn't send its own reply body.
That includes a single headers frame with the end_stream flag (i.e. trailers-only), and two headers frames (headers+trailers).

Comment (ascheglov) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: Removed it.
